src/db.rs

```diff
@@ -1,22 +1,32 @@
-use super::memtable::MemTable;
-use super::{format, Error, Result};
-use crate::wal::Wal;
-use bytes::Bytes;
+mod opt;
+
+use super::memtable::{MemTable, MemTables};
+use super::{Error, Result};
+use crate::entry::Entry;
+use crate::value::{Request, Value};
+
+pub use opt::AgateOptions;
+
 use std::fs;
 use std::path::{Path, PathBuf};
 use std::sync::Arc;
+use std::sync::RwLock;
 
 pub struct Core {
-    wal: Wal,
-    memtable: MemTable,
+    mt: RwLock<MemTables>,
```

Comment:
How about move the lock into `MemTables`? And `Mutex` is probably a better choice.

Reply:
This `RwLock` guards list of `MemTables` (e.g. immutable list). Using `RwLock` ensures that we could parallel fetch data from memtables, while blocking reads only when we're flushing mutable table to immutable one.

And for moving that lock into `MemTables`, I think that is OK. Meanwhile, I would also like to know, is it our design principle to hide locks in structs, and make all functions callable with immutable reference in structs like `MemTable`, `Agate`, `LevelController`, etc.?

Reply:
> If we're using Mutex, we may have to hide all memtables behind an Arc to reduce lock contention when parallel read data. Is it right?

Why? I think the time it consumes is just creating `MemTableView`, which should be very quick. `Mutex` usually has a better performance than `RwLock`, you can bench and see if it's an improvement.

All functions of `Agate` should should take immutable reference and `Agate` should be `Sync + Send`.

Reply:
Yes, thanks for pointing that out. I'll use `Mutex` for `MemTables`.